### PR TITLE
Add cart restrictors

### PR DESCRIPTION
**Pull request by carstendietrich**
The Restriction Service provides a port for implementing product restrictions. By using Dingo multibinding to `cart.MaxQuantityRestrictor`,
you can add your own restriction to the service. The `Restrict` function returns the remaining allowed quantity or `math.MaxInt32` for infinity.

The Service itself consolidates all bound restrictors and returns the most restricting rule (this is the smallest quantity allowed by any restrictor).
If all restrictors return infinity or there is no restrictor bound at all, an error of type `*ErrNoRestriction` is returned.

### CODE REVIEW

**Comment by danielpoe** [comment]
Cool -  I would use the restrictor in the cartService add and update item method

File: cart/domain/cart/services.go [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+import (
+	"context"
+
+	"flamingo.me/flamingo-commerce/v3/product/domain"
+)
+
+type (
+	// MaxQuantityRestrictor returns the maximum qty allowed for a given product and cart
+	MaxQuantityRestrictor interface {
+		// Restrict must return the maximum allowed qty or `int(^uint(0) >> 1)` for infinity
```
Comment (masmrlar) [inline cart/domain/cart/services.go]: Hm, maybe a

```go
math.MaxInt32
```
would also be a bit more readable on this comment line

Comment (carstendietrich) [reply cart/domain/cart/services.go]: Yes, your right. Missed that spot when I switched to the math constant. I'll updated the comment accordingly.

**Comment by masmrlar** [comment]
@carstendietrich can you also resolve the conflict in cart/Readme.md?

**Comment by carstendietrich** [comment]
> @carstendietrich can you also resolve the conflict in cart/Readme.md?

yes, I did a rebase should be fine now 😊

**Comment by carstendietrich** [comment]
@tessig I pushed the requested changes please have a look 😊

File: cart/application/cartService.go [hunk 2]
```diff
@@ -521,6 +545,18 @@ func (cs *CartService) checkProductForAddRequest(ctx context.Context, session *w
 	return addRequest, product, nil
 }
 
+func (cs *CartService) checkProductQtyRestrictions(ctx context.Context, product productDomain.BasicProduct, cart *cartDomain.Cart, qtyToCheck int) error {
+	restrictionResult := cs.restrictionService.RestrictQty(ctx, product, cart)
+
+	if restrictionResult.IsRestricted {
+		if qtyToCheck > restrictionResult.RemainingDifference {
+			return errors.Errorf("Can't update item quantity, product max quantity would be exceeded")
```
Comment (tessig) [inline cart/application/cartService.go]:
```suggestion
			return errors.Errorf("Can't update item quantity, product max quantity of %d would be exceeded", restrictionResult.MaxAllowed)
```

File: cart/application/restrictionService_test.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			rs := &application.RestrictionService{}
+			rs.Inject(tt.fields.qtyRestrictors)
+			got := rs.RestrictQty(tt.args.ctx, tt.args.product, tt.args.cart)
+			if !reflect.DeepEqual(got, tt.expectedRestrictionResult) {
+				t.Errorf("RestrictionService.RestrictQty() got = %v, expected = %v", got, tt.expectedRestrictionResult)
+
+				return
```
Comment (tessig) [inline cart/application/restrictionService_test.go]: I think this is unnecessary here

File: cart/application/cartService.go (hunk 2 above)
Comment (carstendietrich) [reply cart/application/cartService.go]: done.

File: cart/application/restrictionService_test.go (hunk 3 above)
Comment (carstendietrich) [reply cart/application/restrictionService_test.go]: done.

**Pull request merged**